File: android/src/main/java/flutter/curiosity/scan/ScanUtils.java
package flutter.curiosity.scan;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.ContextWrapper;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.util.Log;

import com.google.zxing.BinaryBitmap;
import com.google.zxing.RGBLuminanceSource;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;
import com.google.zxing.qrcode.QRCodeReader;

import java.io.File;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.cert.X509Certificate;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import flutter.curiosity.utils.NativeUtils;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class ScanUtils extends ContextWrapper {

    private QRCodeReader reader = new QRCodeReader();
    private Executor executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler();

    public ScanUtils(Context base) {
        super(base);
    }


    public void scanImagePath(MethodCall call, final MethodChannel.Result result) {
        final String path = call.argument("path");
        final File file = new File(path);
        if (file.isFile()) {
            executor.execute(() -> {
                Bitmap bitmap = BitmapFactory.decodeFile(path);
                scan(bitmap, result);
            });
        } else {
            result.success("");
        }
    }

    public void scan(Bitmap bitmap, final MethodChannel.Result result) {
        int height = bitmap.getHeight();
        int width = bitmap.getWidth();
        try {
            int[] pixels = new int[width * height];
            bitmap.getPixels(pixels, 0, width, 0, 0, width, height);
            RGBLuminanceSource source = new RGBLuminanceSource(
                    width,
                    height, pixels);
            BinaryBitmap binaryBitmap = new BinaryBitmap(new HybridBinarizer(source));
            final Result decode = reader.decode(binaryBitmap, NativeUtils.getHints());
            Log.d("result", "analyze: decode:" + decode.toString());
            handler.post(() -> result.success(NativeUtils.scanDataToMap(decode)));
        } catch (Exception e) {
            Log.d("result", "analyze: error");
            handler.post(() -> result.success(null));
        }

    }

    public void scanImageUrl(MethodCall call, final MethodChannel.Result result) {
        final String url = call.argument("url");
        executor.execute(() -> {
            try {
                URL myUrl = new URL(url);
                Bitmap bitmap;
                assert url != null;
                if (url.startsWith("https")) {
                    HttpsURLConnection connection = (HttpsURLConnection) myUrl.openConnection();
                    connection.setReadTimeout(6 * 60 * 1000);
                    connection.setConnectTimeout(6 * 60 * 1000);
                    TrustManager[] tm = {new MyX509TrustManager()};
                    SSLContext sslContext = SSLContext.getInstance("TLS");
                    sslContext.init(null, tm, new java.security.SecureRandom());
                    // 从上述SSLContext对象中得到SSLSocketFactory对象
                    SSLSocketFactory ssf = sslContext.getSocketFactory();
                    connection.setSSLSocketFactory(ssf);
                    connection.connect();
                    bitmap = BitmapFactory.decodeStream(connection.getInputStream());
                } else {
                    HttpURLConnection connection = (HttpURLConnection) myUrl.openConnection();
                    connection.setReadTimeout(6 * 60 * 1000);
                    connection.setConnectTimeout(6 * 60 * 1000);
                    connection.connect();
                    bitmap = BitmapFactory.decodeStream(connection.getInputStream());
                }
                scan(bitmap, result);


            } catch (Exception e) {
                Log.d("result", "analyze: error");
                handler.post(() -> result.success(null));
            }
        });
    }

    public void scanImageMemory(MethodCall call, final MethodChannel.Result result) {
        final byte[] unit8List = call.argument("unit8List");
        executor.execute(() -> {
            try {
                assert unit8List != null;
                Bitmap bitmap;
                bitmap = BitmapFactory.decodeByteArray(unit8List, 0, unit8List.length);
                scan(bitmap, result);
            } catch (Exception e) {
                Log.d("result", "analyze: error");
                handler.post(() -> result.success(null));
            }
        });
    }

    private static class MyX509TrustManager implements X509TrustManager {

        // 检查客户端证书
        @SuppressLint("TrustAllX509TrustManager")
        public void checkClientTrusted(X509Certificate[] chain, String authType) {
        }

        // 检查服务器端证书
        @SuppressLint("TrustAllX509TrustManager")
        public void checkServerTrusted(X509Certificate[] chain, String authType) {
        }

        // 返回受信任的X509证书数组
        public X509Certificate[] getAcceptedIssuers() {
            return null;
        }
    }
}

File: android/src/main/kotlin/flutter/curiosity/scan/ScanUtils.kt
package flutter.curiosity.scan

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Handler
import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.*
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

object ScanUtils {
    private val executor: Executor = Executors.newSingleThreadExecutor()
    private val handler = Handler()
    private val multiFormatReader: MultiFormatReader = MultiFormatReader()

    fun scanImagePath(call: MethodCall, result: MethodChannel.Result) {
        val path = call.argument<String>("path")
        assert(path != null)
        val file = File(path.toString())
        if (file.isFile) {
            executor.execute {
                val bitmap = BitmapFactory.decodeFile(path)
                handler.post { result.success(scan(bitmap)) }
            }
        } else {
            result.success(null)
        }
    }


    fun scanImageUrl(call: MethodCall, result: MethodChannel.Result) {
        val url = call.argument<String>("url")
        executor.execute {
            try {
                val myUrl = URL(url)
                val bitmap: Bitmap
                assert(url != null)
                if (url!!.startsWith("https")) {
                    val connection = myUrl.openConnection() as HttpsURLConnection
                    connection.readTimeout = 6 * 60 * 1000
                    connection.connectTimeout = 6 * 60 * 1000
                    val tm = arrayOf<TrustManager>(X509Trust())
                    val sslContext = SSLContext.getInstance("TLS")
                    sslContext.init(null, tm, SecureRandom())
                    // 从上述SSLContext对象中得到SSLSocketFactory对象
                    val ssf = sslContext.socketFactory
                    connection.sslSocketFactory = ssf
                    connection.connect()
                    bitmap = BitmapFactory.decodeStream(connection.inputStream)
                } else {
                    val connection = myUrl.openConnection() as HttpURLConnection
                    connection.readTimeout = 6 * 60 * 1000
                    connection.connectTimeout = 6 * 60 * 1000
                    connection.connect()
                    bitmap = BitmapFactory.decodeStream(connection.inputStream)
                }

                handler.post { result.success(scan(bitmap)) }
            } catch (e: Exception) {
                handler.post { result.success(null) }
            }
        }
    }

    fun scanImageMemory(call: MethodCall, result: MethodChannel.Result) {
        val unit8List = call.argument<ByteArray>("unit8List")
        assert(unit8List != null)
        executor.execute {
            try {
                val bitmap: Bitmap = BitmapFactory.decodeByteArray(unit8List, 0, unit8List!!.size)
                handler.post { result.success(scan(bitmap)) }
            } catch (e: Exception) {
                handler.post { result.success(null) }
            }
        }
    }

    private fun scan(bitmap: Bitmap): Map<String, Any>? {
        multiFormatReader.setHints(hints)
        val height = bitmap.height
        val width = bitmap.width
        val pixels = IntArray(width * height)
        return try {
            bitmap.getPixels(pixels, 0, width, 0, 0, width, height)
            val source = RGBLuminanceSource(
                    width,
                    height, pixels)
            val binaryBitmap = BinaryBitmap(HybridBinarizer(source))
            val result = multiFormatReader.decode(binaryBitmap)
            scanDataToMap(result)
        } catch (e: Exception) {
            val data: MutableMap<String, Any> = HashMap()
            data["code"] = "Unrecognized data"
            data["type"] = 0
            data
        }

    }

    private class X509Trust : X509TrustManager {
        // 检查客户端证书
        @SuppressLint("TrustAllX509TrustManager")
        override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {
        }

        // 检查服务器端证书
        @SuppressLint("TrustAllX509TrustManager")
        override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
        }

        // 返回受信任的X509证书数组
        override fun getAcceptedIssuers(): Array<X509Certificate>? {
            return null
        }
    }

    // 这里设置可扫描的类型
    val hints: Map<DecodeHintType, Any>
        get() {
            val decodeFormats: MutableCollection<BarcodeFormat> = ArrayList<BarcodeFormat>()
            //一维码
            decodeFormats.add(BarcodeFormat.UPC_A)
            decodeFormats.add(BarcodeFormat.UPC_E)
            decodeFormats.add(BarcodeFormat.EAN_13)
            decodeFormats.add(BarcodeFormat.EAN_8)
            decodeFormats.add(BarcodeFormat.CODABAR)
            decodeFormats.add(BarcodeFormat.CODE_39)
            decodeFormats.add(BarcodeFormat.CODE_93)
            decodeFormats.add(BarcodeFormat.CODE_128)
            decodeFormats.add(BarcodeFormat.ITF)
            decodeFormats.add(BarcodeFormat.RSS_14)
            decodeFormats.add(BarcodeFormat.RSS_EXPANDED)
            //二维码
            decodeFormats.add(BarcodeFormat.QR_CODE)
            decodeFormats.add(BarcodeFormat.AZTEC)
            decodeFormats.add(BarcodeFormat.DATA_MATRIX)
//            decodeFormats.add(BarcodeFormat.MAXICODE)
//            decodeFormats.add(BarcodeFormat.PDF_417)
            val hints: MutableMap<DecodeHintType, Any> = mutableMapOf()
            hints[DecodeHintType.CHARACTER_SET] = "UTF-8"
            hints[DecodeHintType.POSSIBLE_FORMATS] = decodeFormats
            hints[DecodeHintType.TRY_HARDER] = true
            return hints
        }

    fun scanDataToMap(result: Result?): Map<String, Any>? {
        if (result == null) return null
        val data: MutableMap<String, Any> = HashMap()
        data["code"] = result.text
        data["type"] = result.barcodeFormat.ordinal
        return data
    }
}